internal/refreshtoken/refreshtoken.go

```diff
@@ -0,0 +1,73 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package refreshtoken
+
+import (
+	"bytes"
+	"context"
+	"time"
+
+	"github.com/hashicorp/boundary/internal/errors"
+)
+
+type ResourceType int
+
+const (
+	ResourceTypeUnknown ResourceType = iota
+	ResourceTypeSession
+	ResourceTypeTarget
+)
+
+// GrantsHasher defines the interface used
+// to retrieve a hash of a users grants.
+type GrantsHasher interface {
+	GrantsHash(ctx context.Context) ([]byte, error)
+}
+
+// A RefreshToken is returned in list endpoints for the purposes of pagination
+type RefreshToken struct {
+	CreatedTime         time.Time
+	ResourceType        ResourceType
+	PermissionsHash     []byte
+	LastItemId          string
+	LastItemUpdatedTime time.Time
+}
```

Comment:
I think it might make sense to have a constructor function for this, to ensure it is valid and to help with constructing tokens, ie:

```go
func New(res boundary.Resource, permissionsHash []byte) *RefreshToken {
  return &RefreshToken{
    CreatedTime: time.Now(),
    ResourceType: res.GetType(), // Note this would need to be created, but would probably be useful to be able to get a resource.Type from a boundary.Resource
    PermissionsHash: permissionsHash,
    LastItemId: res.GetPublicId(),
    LastItemUpdatedTime: res.GetUpdateTime(),
  }
}
```

Reply:
There are two primary users of this type:
1. The application layer when converting from the protobuf type to the domain type.
2. The domain layer when constructing a new one from trusted information.
I don't think either of the use cases are helped by a constructor like this, as it doesn't do validation of the values, and if it's just a matter of creating the struct we can do that inline. We also wouldn't want a constructor to perform validation as it isn't always necessary.

Reply:
> I don't think either of the use cases are helped by a constructor like this, as it doesn't do validation of the values, and if it's just a matter of creating the struct we can do that inline. 

The `New` function above ensures that 3 related values, ResoureType, LastItemId, and LastItemUpdatedTime, all come from the same source. Thus the caller does not need to know or enforce the relationship between these fields. I think it also does perform some validation that is helpful, even if this is happening from a trusted source. For example, it ensures that you don't get a RefreshToken with a ResourceType of a Target, with a LastItemId for a Session.

Note: This would only be used when constructing *new* refresh tokens, not by the application layer for transformation.

Reply:
Fair point, we created a `New`, `Refresh` and `RefreshLastItem` set of functions for mutating the refresh token, please take a look!